model/riscv_insts_zvknhab.sail

```diff
@@ -0,0 +1,256 @@
+  then {
+    handle_illegal();
+    RETIRE_FAIL
+  } else {
+    let 'n = num_elem;
+    let 'm = SEW;
+
+    let vm_val  : vector('n, dec, bool)     = read_vmask(num_elem, 0b1, vreg_name("v0"));
+    let vs2_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vs2);
+    let vs1_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vs1);
+    let vd_val  : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vd);
+    result      : vector('n, dec, bits('m)) = undefined;
+    mask        : vector('n, dec, bool)     = undefined;
+
+    (result, mask) = init_masked_result(num_elem, SEW, LMUL_pow, vd_val, vm_val);
+
+    w : vector(20, dec, bits('m)) = undefined;
+
+    eg_len = (unsigned(vl) / 'n);
+    eg_start = (unsigned(vstart) / 'n);
+
+    foreach (i from eg_start to (eg_len - 1)) {
+      assert(0 <= ((i * 4) + 3) & ((i * 4) + 3) < 'n);
+      w[0]  = vd_val[i*4+0];
+      w[1]  = vd_val[i*4+1];
+      w[2]  = vd_val[i*4+2];
+      w[3]  = vd_val[i*4+3];
+
+      w[4]  = vs2_val[i*4+0];
+      w[9]  = vs2_val[i*4+1];
+      w[10] = vs2_val[i*4+2];
+      w[11] = vs2_val[i*4+3];
+
+      w[12] = vs1_val[i*4+0];
+      w[13] = vs1_val[i*4+1];
+      w[14] = vs1_val[i*4+2];
+      w[15] = vs1_val[i*4+3];
+
+      w[16] = sig1(w[14], SEW) + w[9] +  sig0(w[1], SEW) + w[0];
+      w[17] = sig1(w[15], SEW) + w[10] + sig0(w[2], SEW) + w[1];
+      w[18] = sig1(w[16], SEW) + w[11] + sig0(w[3], SEW) + w[2];
+      w[19] = sig1(w[17], SEW) + w[12] + sig0(w[4], SEW) + w[3];
+
+      assert(0 <= 3 & 3 < 'n);
+      result[i*4+0] = w[16];
+      result[i*4+1] = w[17];
+      result[i*4+2] = w[18];
+      result[i*4+3] = w[19];
+    };
+
+    write_single_vreg(num_elem, 'm, vd, result);
+    vstart = EXTZ(0b0);
+    RETIRE_SUCCESS
+  }
+}
+
+/* VSHA2C[HL].VV */
+
+mapping ch_or_cl : string <-> bits(7) = {
+  "ch" <-> 0b1011101,
+  "cl" <-> 0b1011111,
+}
+
+mapping vsha2c_mnemonic : bits(7) <-> string = {
+  0b1011101 <-> "vsha2ch.vv",
+  0b1011111 <-> "vsha2cl.vv",
+}
+
+union clause ast = RISCV_VSHA2c : (regidx, regidx, regidx, string)
+
+mapping clause encdec = RISCV_VSHA2c(vs2, vs1, vd, suffix) if (haveRVV() & (haveZvknha() | haveZvknhb()))
+ <-> ch_or_cl(suffix) @ vs2 @ vs1 @ 0b010 @ vd @ 0b1110111 if (haveRVV() & (haveZvknha() | haveZvknhb()))
+
+mapping clause assembly = RISCV_VSHA2c(vs2, vs1, vd, suffix)
+ <-> vsha2c_mnemonic(ch_or_cl(suffix)) ^ spc() ^ vreg_name(vd)
+				       ^ spc() ^ vreg_name(vs2)
+				       ^ spc() ^ vreg_name(vs1)
+
+function clause execute (RISCV_VSHA2c(vs2, vs1, vd, suffix)) = {
+  let SEW      = get_sew();
+  let LMUL_pow = get_lmul_pow();
+  let LMUL     = if LMUL_pow < 0 then 0 else LMUL_pow;
+  let VLEN     = int_power(2, get_vlen_pow());
+  let num_elem = get_num_elem(LMUL_pow, SEW);
+
+  if (zvk_check_elements(VLEN, num_elem, LMUL, SEW) == false)
+  then {
+    handle_illegal();
+    RETIRE_FAIL
+  } else {
+    let 'n = num_elem;
+    let 'm = SEW;
+    assert('m == 32);
```

Comment:
is this only supporting SHA-256 ? where is tha SHA-512 (SEW=64) support ?

Reply:
Addressed.

Reply:
Does this mean that no SHA-512 tests were executed before or am I missing something ?

Reply:
These have only been tested against QEMU(see https://github.com/riscv-software-src/riscv-isa-sim/pull/1303#issuecomment-1514634288).

Missing SEW=64 occurred due to misreading of the Specification. 

FYI, in SAIL we manually hardcode `VLEN`(wouldn't be long until we will be able to configure it via Command Line options).